proposal/nested-map-functions.md

```diff
@@ -0,0 +1,391 @@
+
+## Summary
+
+> This section is non-normative.
+
+This proposal updates existing map functions with better support for inspection
+and manipulation of nested maps, as well as adding new functions to the
+`sass:map` module. For existing legacy functions (`get()`, `has-key()`,
+`merge()`) the new behavior will be accessible through both the `sass:map`
+module, and global legacy names (`map-get()`, `map-has-key()`, `map-merge()`).
+New functions (`set()`, `deep-merge()`) will only be available inside the
+`sass:map` module.
+
+The `has-key()` and `get()` functions both accept multiple `$keys...`:
+
+```scss
+@use 'sass:map';
+
+$nav: (
+  'bg': 'gray',
+  'color': (
+    'hover': (
+      'search': yellow,
+      'home': red,
+      'filter': blue,
+    ),
+  ),
+);
+
+$has-search: map.has-key($nav, 'color', 'hover', 'search'); // true
+$search-hover: map.get($nav, 'color', 'hover', 'search'); // yellow
+```
+
+The `merge()` function now accepts multiple `$keys...` between the two maps
+being merged. The keys form a path to the nested location in `$map1` where
+`$map2` should be merged. For example, we update the hover colors in our `$nav`
+map above:
+
+```scss
+@use 'sass:map';
+
+$new-hover: (
+  'search': green,
+  'logo': orange,
+);
+
+$nav: map.merge($nav, 'color', 'hover', $new-hover);
+
+// $nav: (
+//   'bg': 'gray',
+//   'color': (
+//     'hover': (
+//       'search': green,
+//       'home': red,
+//       'filter': blue,
+//       'logo': orange,
+//     ),
+//   ),
+// );
+```
+
+This proposal also adds a `set()` function to `sass:map`, with a similar syntax,
+returning a map with any nested key set to a specific value. To achieve the
+same output as our merge example, we can set each key individually:
+
+```scss
+@use 'sass:map';
+
+$nav: map.set($nav, 'color', 'hover', 'search', green);
+$nav: map.set($nav, 'color', 'hover', 'logo', orange);
+```
+
+And finally, a new `deep-merge()` function in the `sass:map` module allows
+merging two or more nested maps. This works much like the existing `merge()`
+function, but when both maps have a nested-map at the same key, those nested
+maps are also merged:
+
+```scss
+@use 'sass:map';
+
+$nav: (
+  'bg': 'gray',
+  'color': (
+    'hover': (
+      'search': yellow,
+      'home': red,
+      'filter': blue,
+    ),
+  ),
+);
+
+$update: (
+  'bg': white,
+  'color': (
+    'hover': (
+      'search': green,
+      'logo': orange,
+    )
+  )
+);
+
+$nav: map.deep-merge($nav, $update);
+
+// $nav: (
+//   'bg': white,
+//   'color': (
+//     'hover': (
+//       'search': green,
+//       'home': red,
+//       'filter': blue,
+//       'logo': orange,
+//     ),
+//   ),
+// );
+```
+
+## Functions
+
+### `get()`
+
+This proposal add a new overload to the existing `get()` function with lower
+priority than the existing signature.
+
+> This means that the new overload is only called if the existing signature
```

Comment:
Does this actually need to be an overload? Wouldn't `get($map, $key)` be covered by the multiple keys version?

Reply:
That's a good point. Updated.